// reaction-base/lib/log/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setLevel = setLevel;

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _loglevel = require('loglevel');

var _loglevel2 = _interopRequireDefault(_loglevel);

var _winston = require('winston');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// eslint-disable-next-line no-new-func
var isBrowser = new Function('try { return this === window } catch (e) { return false }');
var log = createLogger();

function createLogger() {
  if (isBrowser()) {
    _loglevel2.default.setLevel(
    // eslint-disable-next-line no-undef
    _lodash2.default.get(window, 'reaction.config.environment') === 'production' ? 'error' : 'info');
    return _loglevel2.default;
  }
  return new _winston.Logger({
    level: 'info',
    exitOnError: false,
    transports: [new _winston.transports.Console({
      prettyPrint: true,
      colorize: true,
      showLevel: false
    })]
  });
}

function setLevel(level) {
  if (isBrowser()) {
    log.setLevel(level);
  } else {
    log.level = level;
  }
}

exports.default = log;
